plugins/logitech-tap/fu-logitech-tap-sensor-device.c

```diff
@@ -0,0 +1,371 @@
+
+/* enable/disable TDE mode */
+const guchar kHidMcuTdeReportId = 0x1A;
+const guchar kHidMcuTdeModeSelector = 0x02;
+const guchar kHidMcuTdeModeEnable = 0x01;
+const guchar kHidMcuTdeModeDisable = 0x00;
+
+/* serial number of the device */
+const guchar kHidMcuCmdSetSerialNumber = 0x1C;
+const guchar kHidMcuCmdGetSerialNumber = 0x1D;
+const guchar kHidMcuSerialNumberSetReportByte1 = 0x00;
+const guchar kHidMcuSerialNumberSetReportByte2 = 0x70;
+const guchar kHidMcuSerialNumberSetReportByte3 = 0x0E;
+const guchar kHidMcuSerialNumberSetReportByte4 = 0x00;
+
+/* reboot device */
+const guchar kHidReportIdMcuSetCmd = 0x1a;
+
+const guint kLogiDefaultSensorSleepIntervalMs = 50;
+
+struct _FuLogitechTapSensorDevice {
+	FuLogitechTapDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuLogitechTapSensorDevice, fu_logitech_tap_sensor_device, FU_TYPE_LOGITECH_TAP_DEVICE)
+
+static gboolean
+fu_logitech_tap_sensor_device_set_feature(FuLogitechTapSensorDevice *self, const guint8 *data, guint datasz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	fu_dump_raw(G_LOG_DOMAIN, "HidSetFeature", data, datasz);
+	return fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				    HIDIOCSFEATURE(datasz),
+				    (guint8 *)data,
+				    NULL,
+				    FU_LOGITECH_TAP_SENSOR_DEVICE_IOCTL_TIMEOUT,
+				    error);
+#else
+    /* failed */
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_get_feature(FuLogitechTapSensorDevice *self, guint8 *data, guint datasz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	g_autoptr(GError) error_local = NULL;
+	fu_dump_raw(G_LOG_DOMAIN, "HidGetFeatureReq", data, datasz);
+
+	/* try HIDIOCGINPUT request in case of failure */
+	if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCGFEATURE(datasz),
+				  data,
+				  NULL,
+				  FU_LOGITECH_TAP_SENSOR_DEVICE_IOCTL_TIMEOUT,
+				  &error_local)) {
+		g_debug("failed to send get request, retrying: %s", error_local->message);
+		if (!fu_udev_device_ioctl(FU_UDEV_DEVICE(self),
+				  HIDIOCGINPUT(datasz),
+				  data,
+				  NULL,
+				  FU_LOGITECH_TAP_SENSOR_DEVICE_IOCTL_TIMEOUT,
+				  error))
+			return FALSE;
+	}
+
+	fu_dump_raw(G_LOG_DOMAIN, "HidGetFeatureRes", data, datasz);
+
+	return TRUE;
+#else
+	/* failed */
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_logitech_tap_sensor_device_set_tde(FuDevice *device, const guchar tde_mode, GError **error)
+{
+	FuLogitechTapSensorDevice *self = FU_LOGITECH_TAP_SENSOR_DEVICE(device);
+	guint8 set_data[HID_SET_DATA_LEN] = {kHidMcuTdeReportId, kHidMcuTdeModeSelector, tde_mode, 0, 0};
+
+	/* enable/disable TDE mode */
+	if (!fu_logitech_tap_sensor_device_set_feature(self,
+						      (guchar *)set_data,
+							  HID_SET_DATA_LEN,
+						      error))
+	return FALSE;
```

Comment:
whitespace is messed up in this function, both in the error handling and arguments

Reply:
Sorry about white space. Fixed.